zip-0302.rst

```diff
@@ -45,13 +45,27 @@ contents of a memo. It does not define consensus requirements.
      + strip any trailing zero bytes
      + decode it as a UTF-8 string (if decoding fails then report an error).
 
++ If the first byte has a value of 0xF5, then the reader MUST:
+     + Interpret the next few bytes (1 to 9 of them) as a 64-bit unsigned variable-length
+       integer [#Bitcoin-CompactSize]_, and use it as an arbitrary application-defined
+       "type" field.
+     + Interpret the next 1 or 2 bytes as a CompactSize value constrained to the range 0..510, and use
```

Comment:
Strictly speaking this can't be more than 507, because the 0xF5 || type || length is encoded as at least 3 bytes for length <= 252 and 5 bytes for lengths > 252, therefore the maximum is 512 - 5. But it's correct as stated because we only need a conservative upper bound.

Reply:
It’s probably not worth squeezing out the extra couple bytes, but we _could_ store `512 - length` instead of storing `length`, which gets us a range 0..509, so then lengths over 260 only require one byte for the “length”, and shorter messages require three (which they aren’t using anyway).

Reply:
I think the ZIP mixes up the encodings it mentions – the _type_ is stored as CompactSize (1–9 bytes), and the length is stored as VarInt (which fits a value of 508 in 2 bytes, not 3). I found https://bitcoin.stackexchange.com/a/114585, which explains the encodings better than the Bitcoin docs do.

However, just using a regular 2-byte length
* also leaves 508 bytes for the memo;
* is simpler; and
* is consistent with the nested type||length in #247, which says “Total length of encoded multipart memo (2 bytes, as a 16-bit little-endian integer)”.

The multipart length obviously gets no benefit from a variable-length encoding, so I think we should just use a non-variable length 2-byte uint here as well.